### PR TITLE
[GraphBolt] DataLoader end-to-end example

### LINKED ISSUE #5831: Scheduler first E2E version

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This provides an end-to-end DataLoader example using DataPipes.  See the test case for end-to-end usage.

It does subgraph sampling with multiprocessing, and feature fetching with daemonic Python thread.  Please see if you can understand the logic better.

Closes #5831 .

The logic for CUDA stream event recording is from PR #4250.  So I'm requesting the author @yaox12 for review as well to see if my understanding of stream event recording is correct.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/cuda.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+"""Graph Bolt CUDA-related Data Pipelines"""
```
Comment (frozenbugs) [inline python/dgl/graphbolt/cuda.py:1]:
```suggestion
"""GraphBolt CUDA-related Data Pipelines"""
```

File: python/dgl/graphbolt/dataloader.py [hunk 2]
```diff
@@ -1 +1,78 @@
 """Graph Bolt DataLoaders"""
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dataloader.py:1]:
```suggestion
"""GraphBolt DataLoaders"""
```

File: python/dgl/graphbolt/cuda.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    recorded stream event.  In this case, normally :class:`WaitStreamEvent`
+    should follow, which waits for the recorded stream event and yields the data
+    afterwards.
+
+    Parameters
+    ----------
+    dp : DataPipe
+        The DataPipe.
+    device : torch.device
+        The PyTorch CUDA device.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/cuda.py:39]: Does it have to be CUDA device, what about copy to CPU?

File: python/dgl/graphbolt/feature_fetcher.py
Comment (frozenbugs) [inline python/dgl/graphbolt/feature_fetcher.py:1]: Shouldn't we also have ItemLoader(IterDataPipe) to wrap the item loader?

File: python/dgl/graphbolt/feature_fetcher.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+"""Feature fetchers"""
+
+from torchdata.datapipes.iter import IterDataPipe
+
+
+class FeatureFetcher(IterDataPipe):
+    """A mock feature fetcher."""
+
+    def __init__(self, dp, feature_fetch_func):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/feature_fetcher.py:9]:
```suggestion
    def __init__(self, dp, feature_store):
```

Comment (frozenbugs) [reply python/dgl/graphbolt/feature_fetcher.py:9]: The feature fetcher should intake FeatureStore, and rely on the *read* api.

File: python/dgl/graphbolt/subgraph_sampler.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+"""Subgraph samplers"""
+
+from torchdata.datapipes.iter import IterDataPipe
+
+
+class MockSubgraphSampler(IterDataPipe):
+    """A mock subgraph sampler."""
+
+    def __init__(self, dp, graph, sampler):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/subgraph_sampler.py:9]: what's the definition of sampler here?

Comment (frozenbugs) [reply python/dgl/graphbolt/subgraph_sampler.py:9]: It is probably better to pass in a sample_func such that users can define it outside.

File: tests/python/pytorch/graphbolt/test_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import backend as F
+import dgl
+import dgl.graphbolt
+import torch
+
+
+def test_DataLoader():
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_dataloader.py:7]:
```
itemset = torch.arange(N)
minibatch_sampler = dgl.graphbolt.MinibatchSampler(itemset, 4)

graph = (load mock / test csc sampling graph)
def neighbor_sampler(seed_nodes):
  return graph.sample_neighbors(seed_nodes, fanouts=4 ...)
subgraph_sampler = dgl.graphbolt.SubgraphSampler(minibatch_sampler, neighbor_sampler)

feature_store = InMemoryFeatureStore(xxx)
feature_fetcher = dgl.graphbolt.FeatureFetcher(subgraph_sampler, feature_store)

other datapipe.

dataloader = dgl.graphbolt.Dataloader(feature_fetcher ... )
